### PR TITLE
Implement Prism -> Sorbet translation for anonymous Array and Hash splats

### PR DESCRIPTION
**Pull request by amomchilov**
### Motivation
Builds on top of #42 and #225 to add support for anonymous splats.

### Test plan

See included automated tests.

### CODE REVIEW

File: parser/prism/Translator.cc [hunk 1]
```diff
@@ -126,12 +126,8 @@ std::unique_ptr<parser::Node> Translator::translate(pm_node_t *node) {
...
         }
         case PM_ASSOC_SPLAT_NODE: {
-            auto assocSplatNode = reinterpret_cast<pm_assoc_splat_node *>(node);
-            pm_location_t *loc = &assocSplatNode->base.location;
-
-            auto value = translate(assocSplatNode->value);
-
-            return make_unique<parser::Kwsplat>(parser.translateLocation(loc), std::move(value));
+            unreachable("PM_ASSOC_SPLAT_NODE is handled separately in `Translator::translateHash()`, because its "
+                        "translation depends on whether its used in a hash or in a method call");
```
Comment (amomchilov) [inline parser/prism/Translator.cc:130]: This prevents us from needing to maintain a parsing context, that remembers whether this recursive call to `Translator::translate()` originated from a `PM_HASH_NODE` or a `PM_KEYWORD_HASH_NODE`.

File: parser/prism/Translator.cc [hunk 2]
```diff
@@ -908,7 +907,7 @@ parser::NodeVec Translator::translateMulti(pm_node_list nodeList) {
     parser::NodeVec result;
 
     // Pre-allocate the exactly capacity we're going to need, to prevent growth reallocations.
-    result.reserve(std::distance(prismNodes.begin(), prismNodes.end()));
+    result.reserve(prismNodes.size());
```
Comment (amomchilov) [inline parser/prism/Translator.cc:910]: Overlooked this when I first wrote it :)

File: parser/prism/Translator.cc [hunk 3]
```diff
@@ -956,7 +955,30 @@ std::unique_ptr<parser::Hash> Translator::translateHash(pm_node_t *node, pm_node
...
+            pm_location_t *loc = &prismSplatNode->base.location;
+            auto value = translate(prismSplatNode->value);
+
+            if (value == nullptr) { // An anonymous splat like `f(**)`
+                auto sorbetSplatNode = make_unique<parser::ForwardedKwrestArg>(parser.translateLocation(loc));
+                sorbetElements.emplace_back(std::move(sorbetSplatNode));
+            } else { // Splatting an expression like `f(**h)`
+                auto sorbetSplatNode = make_unique<parser::Kwsplat>(parser.translateLocation(loc), std::move(value));
+                sorbetElements.emplace_back(std::move(sorbetSplatNode));
+            }
```
Comment (amomchilov) [inline parser/prism/Translator.cc:975]: Ideally, I would have wanted something more like this:

```suggestion
            std::unique_ptr<parser::Node> sorbetSplatNode;
            
            if (value == nullptr) { // An anonymous splat like `f(**)`
               sorbetSplatNode = make_unique<parser::ForwardedKwrestArg>(parser.translateLocation(loc));
            } else { // Splatting an expression like `f(**h)`
               sorbetSplatNode = make_unique<parser::Kwsplat>(parser.translateLocation(loc), std::move(value));
            }
            
            sorbetElements.emplace_back(std::move(sorbetSplatNode));
```

But remarkably, there's no easy way to cast a `std::unique_ptr<parser::ForwardedKwrestArg>` to `std::unique_ptr<parser::Node>`, like you can with "regular" C pointers. You'd have add your own helper, like this [`static_unique_ptr_cast()`](https://stackoverflow.com/a/33649683).

Comment (egiurleo) [reply parser/prism/Translator.cc:975]: That's so weird. I feel like we do something similar in the `translateAssignment` template, but maybe there's some subtlety there that makes it different? https://github.com/Shopify/sorbet/blob/prism/parser/prism/Translator.cc#L30

**Review by egiurleo — APPROVED** [review]
Looks great!

File: parser/prism/Translator.cc (hunk 3 above)
Comment (amomchilov) [reply parser/prism/Translator.cc:975]: Huh, I tried this again, exactly as my suggestion said, and it worked perfectly. 🤔

**Pull request merged**